src/ucp/core/ucp_worker.c

```diff
@@ -2360,3 +2390,180 @@ void ucp_worker_print_info(ucp_worker_h worker, FILE *stream)
+}
+
+static unsigned ucp_worker_discard_uct_ep_progress(void *arg)
+{
+    ucp_request_t *req  = (ucp_request_t*)arg;
+    ucs_status_t status = ucp_worker_discard_uct_ep_pending_cb(&req->send.uct);
+
+    return !UCS_STATUS_IS_ERR(status);
+}
+
+static uct_ep_h
+ucp_worker_discard_wireup_ep(ucp_worker_h worker,
+                             ucp_wireup_ep_t *wireup_ep,
+                             unsigned ep_flush_flags,
+                             uct_pending_purge_callback_t purge_cb,
+                             void *purge_arg)
+{
+    uct_ep_h uct_ep;
+    int is_owner;
+
+    ucs_assert(wireup_ep != NULL);
+    ucs_assert(purge_cb != NULL);
+
+    uct_ep_pending_purge(&wireup_ep->super.super, purge_cb, purge_arg);
+
+    if (wireup_ep->aux_ep != NULL) {
+        /* make sure that there is no WIREUP MSGs anymore */
+        uct_ep_pending_purge(wireup_ep->aux_ep,
+                             (uct_pending_purge_callback_t)
+                             ucs_empty_function_do_assert,
+                             NULL);
+
+        /* discard the WIREUP EP's auxiliary EP */
+        ucp_worker_discard_uct_ep(worker, wireup_ep->aux_ep,
+                                  ep_flush_flags,
+                                  purge_cb, purge_arg);
+        ucp_wireup_ep_disown(&wireup_ep->super.super, wireup_ep->aux_ep);
+    }
+
+    is_owner = wireup_ep->super.is_owner;
+    uct_ep   = ucp_wireup_ep_extract_next_ep(&wireup_ep->super.super);
+
+    /* destroy WIREUP EP allocated for this UCT EP, since discard operation
+     * most likely won't have an access to UCP EP as it could be destroyed
+     * by the caller */
+    uct_ep_destroy(&wireup_ep->super.super);
+
+    /* do nothing, if this wireup EP is not an owner for UCT EP */
+    return is_owner ? uct_ep : NULL;
+}
+
+/* must be called with async lock held */
+void ucp_worker_discard_uct_ep(ucp_worker_h worker, uct_ep_h uct_ep,
+                               unsigned ep_flush_flags,
+                               uct_pending_purge_callback_t purge_cb,
+                               void *purge_arg)
+{
+    uct_worker_cb_id_t cb_id = UCS_CALLBACKQ_ID_NULL;
+    ucp_request_t *req;
+    khiter_t UCS_V_UNUSED iter;
+    int ret;
+
+    ucs_assert(uct_ep != NULL);
+
+    if (ucp_wireup_ep_test(uct_ep)) {
+        uct_ep = ucp_worker_discard_wireup_ep(worker, ucp_wireup_ep(uct_ep),
+                                              ep_flush_flags,
+                                              purge_cb, purge_arg);
+        if (uct_ep == NULL) {
+            return;
+        }
+    } else if (purge_cb != NULL) {
+        uct_ep_pending_purge(uct_ep, purge_cb, purge_arg);
+    }
+
+    req = ucp_request_get(worker);
+    if (ucs_unlikely(req == NULL)) {
+        ucs_error("unable to allocate request for discarding UCT EP %p "
+                  "on UCP worker %p", uct_ep, worker);
+        return;
+    }
+
+    ++worker->flush_ops_count;
+    iter = kh_put(ucp_worker_discard_uct_ep_hash,
```

Comment:
minor: maybe just `kh_put(ucp_worker_discard_uct_ep_hash,...`? Then no need in unused `iter` var

Reply:
done